### PR TITLE
DAO3

### PR DESCRIPTION
**Pull request by yurwing**
created model for Car, Driver and Manufacturer
created Storage for manufacturers
created dao and service for Manufacturer
created annotation Dao, Inject and Service

### CODE REVIEW

File: src/main/java/core/basesyntax/Main.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    private static final Injector injector = Injector.getInstance("core.basesyntax");
+
+    public static void main(String[] args) {
+        ManufacturerService manufacturerService = (ManufacturerService)
+                injector.getInstance(ManufacturerService.class);
+        manufacturerService.create(new Manufacturer("Honda", "Japan"));
+        manufacturerService.create(new Manufacturer("Hyundai", "South Korea"));
+        manufacturerService.create(new Manufacturer("Toyota", "Japan"));
+        System.out.println(manufacturerService.getAll());
+        System.out.println(manufacturerService.get(3L));
```
Comment (Baldinini) [inline src/main/java/core/basesyntax/Main.java]: create constant

File: src/main/java/core/basesyntax/Main.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        manufacturerService.delete(3L);
+        System.out.println("After deleting");
+        System.out.println(manufacturerService.getAll());
+        Manufacturer tesla = manufacturerService.create(new Manufacturer("Tesla", "USA"));
+        System.out.println(manufacturerService.getAll());
+        tesla.setName("new Tesla");
+        manufacturerService.update(tesla);
+        System.out.println("After updating");
+        System.out.println(manufacturerService.getAll());
+        manufacturerService.delete(1L);
```
Comment (Baldinini) [inline src/main/java/core/basesyntax/Main.java]: same

File: src/main/java/core/basesyntax/dao/ManufacturerDaoImpl.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    @Override
+    public List<Manufacturer> getAll() {
+        return Storage.manufacturers;
+    }
+
+    @Override
+    public Manufacturer update(Manufacturer manufacturer) {
+        Manufacturer newManufacturer = getAll().stream()
```
Comment (Baldinini) [inline src/main/java/core/basesyntax/dao/ManufacturerDaoImpl.java]:
```suggestion
        Manufacturer newManufacturer = get(manufacturer.getId()).get();
```

**Review by Baldinini — CHANGES_REQUESTED** [review]

**Review by Baldinini — APPROVED** [review]

File: src/main/java/core/basesyntax/dao/ManufacturerDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    public List<Manufacturer> getAll() {
+        return Storage.manufacturers;
+    }
+
+    @Override
+    public Manufacturer update(Manufacturer manufacturer) {
+        Manufacturer newManufacturer = get(manufacturer.getId()).get();
+        newManufacturer.setName(manufacturer.getName());
+        newManufacturer.setCountry(manufacturer.getCountry());
+        return newManufacturer;
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/dao/ManufacturerDaoImpl.java]: we should iterate the list to find the needed object, current approach won't work correctly when you delete element from list and then try to update it

File: src/main/java/core/basesyntax/dao/ManufacturerDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    @Override
+    public boolean delete(Long id) {
+        return delete(get(id).get());
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/dao/ManufacturerDaoImpl.java]: use removeIf instead here that will be more reliable

File: src/main/java/core/basesyntax/model/Car.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+import java.util.List;
+import java.util.Objects;
+
+public class Car {
+    private Long id;
+    private String model;
+    private Manufacturer manufacturer;
+    private List<Driver> drivers;
+
+    public Car(Long id, String model, Manufacturer manufacturer, List<Driver> drivers) {
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/model/Car.java]: checklist

**Review by RomanDubovskyi — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/dao/ManufacturerDaoImpl.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+                .filter(i -> Storage.manufacturers.get(i).getId().equals(manufacturer.getId()))
+                .findFirst()
+                .getAsInt();
+        Storage.manufacturers.set(index, manufacturer);
+        return manufacturer;
+    }
+
+    @Override
+    public boolean delete(Long id) {
+        return Storage.manufacturers.remove(get(id).get());
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/dao/ManufacturerDaoImpl.java]:
```suggestion
        return Storage.manufacturers.removeIf(m -> m.getId().equals(id));
```
and we won't need to use get method inside

File: src/main/java/core/basesyntax/dao/ManufacturerDaoImpl.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        return Storage.manufacturers;
+    }
+
+    @Override
+    public Manufacturer update(Manufacturer manufacturer) {
+        int index = IntStream.range(0, Storage.manufacturers.size())
+                .filter(i -> Storage.manufacturers.get(i).getId().equals(manufacturer.getId()))
+                .findFirst()
+                .getAsInt();
+        Storage.manufacturers.set(index, manufacturer);
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/dao/ManufacturerDaoImpl.java]:
```suggestion
                .forEach(i -> Storage.manufacturers.set(i, manufacturer);
```

File: src/main/java/core/basesyntax/model/Car.java (hunk 6 above)
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/model/Car.java]: what about this comment?

**Review by RomanDubovskyi — CHANGES_REQUESTED** [review]

**Review by RomanDubovskyi — APPROVED** [review]

**Pull request merged**